### ISSUE TITLE
AttributeError: '_ResultMetric' object has no attribute '_forward_pre_hooks'

### ISSUE THREAD

**Issue by jerome-habana** [Lightning-AI/lightning#18123]
### Bug description

using self.log with torch compile results in failure.
https://github.com/pytorch/pytorch/pull/103621 applied to mitigate "failed to reach fixed point" error with python 3.8

Pytorch lightning 2.0.5 used for the experiment

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
def test_compiled_model_to_log_metric_with_cpu(tmp_path):
    class MyModel(BoringModel):
        def training_step(self, batch, batch_idx):
            loss = self.step(batch)
            self.log("loss", loss)
            return loss

    model = MyModel()
    compiled_model = torch.compile(model)

    trainer = Trainer(
        default_root_dir=tmp_path,
        accelerator="cpu",
        fast_dev_run=True,
        devices=1,
        enable_checkpointing=False,
        enable_model_summary=False,
        enable_progress_bar=False,
    )
    trainer.fit(compiled_model)

    assert set(trainer.callback_metrics) == {"loss"}
```

### Error messages and logs

```
>       raise AttributeError("'{}' object has no attribute '{}'".format(
            type(self).__name__, name))
E       AttributeError: '_ResultMetric' object has no attribute '_forward_pre_hooks'
E
E       from user code:
E          File "/home/janand/anaconda3/envs/pylight/lib/python3.8/site-packages/lightning/pytorch/trainer/connectors/logger_connector/result.py", line 188, in __init__
E           super().__init__()
E
E       Set torch._dynamo.config.verbose=True for more information
E
E
E       You can suppress this exception and fall back to eager by setting:
E           torch._dynamo.config.suppress_errors = True

../../../../anaconda3/envs/pylight/lib/python3.8/site-packages/torch/nn/modules/module.py:1617: AttributeError


```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @carmocca @Blaizzy

**Comment by jerome-habana** [comment]
cc @Borda

**Comment by jerome-habana** [comment]
@Borda @carmocca Any updates to this issue ?

**Comment by awaelchli** [comment]
@jerome-habana Thanks for providing the code. I don't know how Lightning can do anything about this to work around the bug in PyTorch. I see that the linked issue has been merged, and I can confirm that running your example on Python 3.9+ has no issues.

**Comment by jerome-habana** [comment]
> @jerome-habana Thanks for providing the code. I don't know how Lightning can do anything about this to work around the bug in PyTorch. I see that the linked issue has been merged, and I can confirm that running your example on Python 3.9+ has no issues.

@awaelchli I do see self.log working well with nn.module and issue seen with lightning module. Do you think its still related to torch issue ?

**Comment by jerome-habana** [comment]
@awaelchli the issue seems to be driven from 'setattr'. I see the result metric can conflict with user defined 'setattr' resulting in the above failure. 
"@wraps(torch.nn.Module.__setattr__)"

**Comment by awaelchli** [comment]
I don't understand what you mean. Are you referring to this setattr?
https://github.com/Lightning-AI/lightning/blob/57ef88f8fccfb05b84b069e428a5d5d5a5ac5a46/src/lightning/pytorch/trainer/connectors/logger_connector/result.py#L281-L283

This is an internal class, so I don't understand what you mean by "user defined".

**Comment by jerome-habana** [comment]
Yep. You can try out the below example

```
import torch
from torch.nn import functional as F
from lightning.pytorch import LightningModule, Trainer
from lightning.pytorch.demos.mnist_datamodule import MNISTDataModule
from torchmetrics import Metric
from functools import wraps
from typing import Any, Union, Optional
```

```
class LitClassifier(LightningModule):
    def __init__(self):
        super().__init__()
        self.l1 = torch.nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

    def training_step(self, batch, batch_idx):
        x, y = batch
        loss = F.cross_entropy(self(x), y)
        self.log("loss", loss)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.02)

```
```
def test_compiled_model_to_log_metric_with_cpu():
    #torch._dynamo.config.verbose=True
    dm = MNISTDataModule(batch_size=32)
    model = LitClassifier()
    compiled_model = torch.compile(model)
    trainer = Trainer(
        accelerator="cpu",
        fast_dev_run=True,
        devices=1,
        enable_checkpointing=False,
        enable_model_summary=False,
        enable_progress_bar=False,
    )
    trainer.fit(compiled_model, datamodule=dm)
    assert set(trainer.callback_metrics) == {"loss"}

def overwrite_torch_functions():
    # wrap torch.nn.Module.__setattr__
    module_set_attr_orig = torch.nn.Module.__setattr__

    @wraps(torch.nn.Module.__setattr__)
    def wrap_set_attr(self, name: str, value: Union[torch.Tensor, 'torch.nn.Module']) -> None:
        if isinstance(value, torch.nn.Module):# and not isinstance(value, Metric):
            try:
                print("------value :: ", value)
            except (RuntimeError):
                pass
        module_set_attr_orig(self, name, value)

    torch.nn.Module.__setattr__ = wrap_set_attr

```   
```
if __name__ == '__main__':
    overwrite_torch_functions()
    test_compiled_model_to_log_metric_with_cpu()

```

**Comment by awaelchli** [comment]
I was able to reduce it to just a torchmetrics example where an equivalent error is raised (the cause of the same error we are seeing in Lightnings _ResultMetric that inherits from torchmetrics):

```py
import torch
import torch.nn as nn
from torchmetrics import Accuracy


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(2, 2)

    def forward(self, x):
        y = Accuracy(task="binary")
        return self.layer(x)


def overwrite_torch_functions():
    module_set_attr_orig = torch.nn.Module.__setattr__

    def wrap_set_attr(self, name, value):
        if isinstance(value, torch.nn.Module):
            print(value)  # <-- calls `__repr__` on the module
        module_set_attr_orig(self, name, value)

    torch.nn.Module.__setattr__ = wrap_set_attr

overwrite_torch_functions()

model = Model()
model = torch.compile(model)
model(torch.rand(2, 2))
```

The instantiation of the metric in the middle of `forward()` [simulates what is happening in Lightning when `self.log()` is called.](https://github.com/Lightning-AI/lightning/blob/6df43685ee2f2dd0c53eefec295e0d6c79796fd2/src/lightning/pytorch/trainer/connectors/logger_connector/result.py#L402-L409)

During the execution of this, the torch-compiled code is calling setattr *during the creation* of that module. This then goes through the patched function here where `print(value)` is evaluated, which in return calls `repr(value)` which accesses attributes from that object that don't exist yet, because the object is being created. 

This recursive dependency is the problem, but I don't know how to solve this. The next step is to reduce this even further to a raw PyTorch example, and then check whether PyTorch can address this use case.

**Comment by Borda** [comment]
cc: @SkafteNicki @justusschock

**Comment by SkafteNicki** [comment]
@awaelchli, just checked and replacing the metric with the most simple module possible, still reproduces the error:
```python
import torch
import torch.nn as nn

class DummyModule(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return x

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(2, 2)

    def forward(self, x):
        y = DummyModule()
        return self.layer(x)

def overwrite_torch_functions():
    module_set_attr_orig = torch.nn.Module.__setattr__

    def wrap_set_attr(self, name, value):
        if isinstance(value, torch.nn.Module):
            print(value)  # <-- calls `__repr__` on the module
        module_set_attr_orig(self, name, value)

    torch.nn.Module.__setattr__ = wrap_set_attr

overwrite_torch_functions()

model = Model()
model = torch.compile(model)
model(torch.rand(2, 2))
```

**Comment by awaelchli** [comment]
@SkafteNicki Thanks for taking a look. This is the first thing I did too, but on torch 2.1 nightly, this seems to work, yet somehow the torchmetrics example does not. I just checked again and you are right, it does show the same error on torch 2.0. So something must have been changed between torch 2.0 and 2.1 that partially fixes this. 

I'll look further what the difference between the raw module and the Metric is. 
I don't expect TorchMetrics to be at fault here, after all, torch.compile should work as long as the Python code is valid.

**Comment by awaelchli** [comment]
I found the difference and extracted a raw PyTorch example:

```py
import torch
import torch.nn as nn

class BinaryStatScores(nn.Module):
    def __init__(self):
        super().__init__()


class Accuracy(nn.Module):
    def __new__(cls):
        return BinaryStatScores()


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(2, 2)

    def forward(self, x):
        y = Accuracy()
        return self.layer(x)


def overwrite_torch_functions():
    module_set_attr_orig = torch.nn.Module.__setattr__

    def wrap_set_attr(self, name, value):
        if isinstance(value, torch.nn.Module):
            print(value)  # <-- calls `__repr__` on the module
        module_set_attr_orig(self, name, value)

    torch.nn.Module.__setattr__ = wrap_set_attr

overwrite_torch_functions()

model = Model()
model = torch.compile(model)
model(torch.rand(2, 2))
```
`torch==2.1.0.dev20230818+cu118`

We can report this to torch and see if it can be addressed. However, this use case of patching out setattr on all nn.Modules is quite exotic and probably not future-proof. @jerome-habana my recommendation for you is to just make the setattr compatible with `torch.compile` explicitly by making it aware of the internal root module. For example, you could handle it like this in your setattr override:

```py
if isinstance(value, torch.nn.Module) and not isinstance(self, torch._dynamo.output_graph.FakeRootModule):
    print(value)
```
This way, you avoid calling print on the dynamo module which resulted in the repr issue.

**Comment by jerome-habana** [comment]
Thanks @awaelchli . As of now, am making it compatible with metric as shown in the example. But your suggestion is better when compared to just metric.

**Comment by awaelchli** [comment]
@jerome-habana Sounds good! Closing the issue since we were able to show this to be a limitation in torch and needs to be addressed there.

**Comment by nathanpainchaud** [comment]
I'm replying on this issue because I'm experiencing a very similar issue, only with up-to-date PyTorch and Lightning versions. I don't know if this is indicative of a regression in PyTorch (or torchmetrics?).

Based off the test examples above, the code below fails for me:
```python
import torch
import torch.nn as nn
from torchmetrics import Accuracy

class DummyModule(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        return x

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(2, 2)

    def forward(self, x):
        # y = DummyModule() # <-- replacing the torchmetric with a dummy module fixes the issue
        y = Accuracy(task="binary")  # <-- this fails
        return self.layer(x)

def overwrite_torch_functions():
    module_set_attr_orig = torch.nn.Module.__setattr__

    def wrap_set_attr(self, name, value):
        if isinstance(value, torch.nn.Module):
            print(value)  # <-- calls `__repr__` on the module
        module_set_attr_orig(self, name, value)

    torch.nn.Module.__setattr__ = wrap_set_attr

if __name__ == '__main__':
    overwrite_torch_functions()

    model = Model()
    model = torch.compile(model)
    model(torch.rand(2, 2))
```

with the following traceback:
```
Traceback (most recent call last):
  File "/data/Work/git/lightning-hydra-template/scripts/test.py", line 37, in <module>
    model(torch.rand(2, 2))
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 465, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/scripts/test.py", line 17, in forward
    def forward(self, x):
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torchmetrics/classification/accuracy.py", line 515, in __new__
    return BinaryAccuracy(threshold, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torchmetrics/classification/stat_scores.py", line 173, in __init__
    super(_AbstractStatScores, self).__init__(**kwargs)
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torchmetrics/metric.py", line 115, in __init__
    self._TORCH_GREATER_EQUAL_2_1 = bool(_TORCH_GREATER_EQUAL_2_1)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torchmetrics/metric.py", line 812, in __setattr__
    super().__setattr__(name, value)
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 1269, in __call__
    return self._torchdynamo_orig_callable(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 1064, in __call__
    result = self._inner_convert(
             ^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 526, in __call__
    return _compile(
           ^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 952, in _compile
    raise InternalTorchDynamoError(
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 924, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 666, in compile_inner
    return _compile_inner(code, one_graph, hooks, transform)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_utils_internal.py", line 87, in wrapper_function
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 699, in _compile_inner
    out_code = transform_code_object(code, transform)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/bytecode_transformation.py", line 1322, in transform_code_object
    transformations(instructions, code_options)
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 219, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/convert_frame.py", line 634, in transform
    tracer.run()
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 2796, in run
    super().run()
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 983, in run
    while self.step():
          ^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 895, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 582, in wrapper
    return inner_fn(self, inst)
           ^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 2279, in CALL
    self._call(inst)
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 2273, in _call
    self.call_function(fn, args, kwargs)
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 830, in call_function
    self.push(fn.call_function(self, args, kwargs))  # type: ignore[arg-type]
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/lazy.py", line 156, in realize_and_forward
    return getattr(self.realize(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/functions.py", line 324, in call_function
    return super().call_function(tx, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/functions.py", line 111, in call_function
    return tx.inline_user_function_return(self, [*self.self_args(), *args], kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 836, in inline_user_function_return
    return InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 3011, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/symbolic_convert.py", line 3077, in inline_call_
    and isinstance(
        ^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/base.py", line 110, in __instancecheck__
    instance = instance.realize()
               ^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/lazy.py", line 63, in realize
    self._cache.realize()
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/lazy.py", line 29, in realize
    self.vt = VariableBuilder(tx, self.source)(self.value)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/builder.py", line 377, in __call__
    vt = self._wrap(value)
         ^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/builder.py", line 651, in _wrap
    return self.wrap_module(value)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/_dynamo/variables/builder.py", line 1379, in wrap_module
    for p in value.parameters():
             ^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 2633, in parameters
    for name, param in self.named_parameters(recurse=recurse):
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 2666, in named_parameters
    yield from gen
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 2601, in _named_members
    for module_prefix, module in modules:
                                 ^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 2815, in named_modules
    if self not in memo:
       ^^^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torchmetrics/metric.py", line 1028, in __hash__
    for key in self._defaults:
               ^^^^^^^^^^^^^^
  File "/data/Work/git/lightning-hydra-template/.venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1931, in __getattr__
    raise AttributeError(
torch._dynamo.exc.InternalTorchDynamoError: AttributeError: 'BinaryAccuracy' object has no attribute '_defaults'

from user code:
   File "/data/Work/git/lightning-hydra-template/scripts/test.py", line 28, in wrap_set_attr
    module_set_attr_orig(self, name, value)

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True

```

The initial issue was linked to dynamo on pytorch's side not supporting `setattr`. However, as far as I could tell this support landed back in PyTorch 2.2. See [PR that fixed the issue](https://github.com/pytorch/pytorch/pull/108549).

I found a [related issue on torchmetric's side](https://github.com/Lightning-AI/torchmetrics/issues/1780), but once again it seemed like it should have been fixed in current PyTorch releases.

Versions in the setup I'm testing on (CPU only):
```txt
torch==2.5.1
lightning==2.4.0
```

I would be happy with any help/suggestion to understand why the example is not working with up-to-date PyTorch and Lightning :slightly_smiling_face:
